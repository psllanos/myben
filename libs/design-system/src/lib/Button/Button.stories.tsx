import type { Story, Meta } from '@storybook/react'
import type { ButtonProps } from './Button'

import Button from './Button'

export default {
    title: 'Components/Button',
    component: Button,
    parameters: { controls: { exclude: ['as', 'className', 'onClick'] } },
    argTypes: {
        children: {
            control: 'text',
            description: 'Button content',
        },
    },
    args: {
        children: 'Click Me!',
        variant: 'primary',
    },
} as Meta

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Base = Template.bind({})
