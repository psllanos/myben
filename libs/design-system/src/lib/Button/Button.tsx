import type { PropsWithChildren, ReactNode } from 'react'
import React from 'react'
import classNames from 'classnames'

const ButtonVariants = Object.freeze({
    primary:
        'px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan',
    secondary:
        'px-4 py-2 rounded text-base bg-gray-500 text-gray-25 shadow hover:bg-gray-400 focus:bg-gray-400 focus:ring-gray-400',
    input: 'px-4 py-2 rounded text-base bg-transparent text-gray-25 border border-gray-200 shadow focus:bg-gray-500 focus:ring-gray-400',
    link: 'px-4 py-2 rounded text-base text-cyan hover:text-cyan-400 focus:text-cyan-300 focus:ring-cyan',
    icon: 'p-0 w-8 h-8 rounded text-2xl text-gray-25 hover:bg-gray-300 focus:bg-gray-200 focus:ring-gray-400',
    profileIcon: 'p-0 w-12 h-12 rounded text-2xl text-gray-25',
    danger: 'px-4 py-2 rounded text-base bg-red text-gray-700 shadow hover:bg-red-400 focus:bg-red-400 focus:ring-red',
    warn: 'px-4 py-2 rounded text-base bg-gray-500 text-red-500 shadow hover:bg-gray-400 focus:bg-gray-400 focus:ring-red',
})

export type ButtonVariant = keyof typeof ButtonVariants

export type ButtonProps = Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'as'> &
    PropsWithChildren<{
        variant?: ButtonVariant

        onClick?: () => void

        disabled?: boolean

        /** Display as a full-width block */
        fullWidth?: boolean

        /** Will use an 'a' tag if specified */
        href?: string

        /** 'a' download attribute */
        download?: string

        /** 'a' target attribute */
        target?: string

        /** Element to use (default is <button> or <a> depending on props) */
        as?: React.ElementType

        className?: string

        leftIcon?: ReactNode

        /** Display spinner based on the value passed */
        isLoading?: boolean
    }>

function Button(
    {
        variant = 'primary',
        fullWidth = false,
        href,
        as,
        children,
        className,
        leftIcon,
        isLoading = false,
        ...rest
    }: ButtonProps,
    ref: React.Ref<HTMLElement>
): JSX.Element {
    const combinedClassName = classNames(
        className,
        ButtonVariants[variant],
        fullWidth && 'w-full',
        'inline-flex items-center justify-center text-center',
        'font-medium leading-6 whitespace-nowrap select-none cursor-pointer',
        'focus:outline-none focus:ring focus:ring-opacity-60',
        'disabled:opacity-50 disabled:pointer-events-none',
        'transition-colors duration-200'
    )

    const Tag = as || (href ? 'a' : 'button')

    return (
        <Tag
            ref={ref}
            className={combinedClassName}
            role="button"
            {...(Tag === 'button' && { type: 'button' })}
            {...(href && { href })} // Adds href only if exists.
            {...rest}
        >
            {leftIcon && <span className="mr-1.5">{leftIcon}</span>}
            {children}
            {isLoading && Tag === 'button' && (
                <span className="ml-1.5">
                    <Spinner fill={variant === 'primary' ? '#16161A' : '#FFFFFF'} />
                </span>
            )}
        </Tag>
    )
}

// Add a spinner to Button component and toggle it via flag

interface SpinnerProps {
    fill: string
}

function Spinner({ fill }: SpinnerProps): JSX.Element {
    return (
        <svg width="20" height="20" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
            <path
                d="M12,1A11,11,0,1,0,23,12,11,11,0,0,0,12,1Zm0,19a8,8,0,1,1,8-8A8,8,0,0,1,12,20Z"
                opacity=".25"
                fill={fill}
            />
            <path
                d="M10.72,19.9a8,8,0,0,1-6.5-9.79A7.77,7.77,0,0,1,10.4,4.16a8,8,0,0,1,9.49,6.52A1.54,1.54,0,0,0,21.38,12h.13a1.37,1.37,0,0,0,1.38-1.54,11,11,0,1,0-12.7,12.39A1.54,1.54,0,0,0,12,21.34h0A1.47,1.47,0,0,0,10.72,19.9Z"
                fill={fill}
            >
                <animateTransform
                    attributeName="transform"
                    type="rotate"
                    dur="0.75s"
                    values="0 12 12;360 12 12"
                    repeatCount="indefinite"
                />
            </path>
        </svg>
    )
}

export default React.forwardRef(Button)
