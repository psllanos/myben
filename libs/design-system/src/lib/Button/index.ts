export { default as Button } from './Button'
export type { ButtonVariant, ButtonProps } from './Button'
