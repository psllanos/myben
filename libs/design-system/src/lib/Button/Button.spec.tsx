import { fireEvent, render, screen } from '@testing-library/react'
import Button from './Button'

describe('Button', () => {
    describe('when rendered with text', () => {
        it('should display the text', () => {
            const component = render(<Button>Hello, World!</Button>)

            expect(screen.getByText('Hello, World!')).toBeInTheDocument()
            expect(component).toMatchSnapshot()
        })
    })

    describe('when pressed', () => {
        it('should call the `onClick` callback', () => {
            const onClickMock = jest.fn()
            render(<Button onClick={onClickMock}>Hello, World!</Button>)

            fireEvent.click(screen.getByRole('button'))
            expect(onClickMock).toHaveBeenCalledTimes(1)
        })
    })

    describe('when passed an `href` value', () => {
        it('should render as an <a> element', () => {
            const component = render(<Button href="https://example.com">Hello, World!</Button>)
            // Ideally we'd actually assert the tag name here, but I can't see a good way to do that
            expect(component).toMatchSnapshot()
        })
    })

    describe('when passed an `as` prop', () => {
        it('should render as the specified element', () => {
            const component = render(<Button as="div">Hello, World!</Button>)
            // Ideally we'd actually assert the tag name here, but I can't see a good way to do that
            expect(component).toMatchSnapshot()
        })
    })
})
