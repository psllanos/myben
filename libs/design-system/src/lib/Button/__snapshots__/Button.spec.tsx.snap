// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when passed an \`as\` prop should render as the specified element 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan inline-flex items-center justify-center text-center font-medium leading-6 whitespace-nowrap select-none cursor-pointer focus:outline-none focus:ring focus:ring-opacity-60 disabled:opacity-50 disabled:pointer-events-none transition-colors duration-200"
        role="button"
      >
        Hello, World!
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan inline-flex items-center justify-center text-center font-medium leading-6 whitespace-nowrap select-none cursor-pointer focus:outline-none focus:ring focus:ring-opacity-60 disabled:opacity-50 disabled:pointer-events-none transition-colors duration-200"
      role="button"
    >
      Hello, World!
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button when passed an \`href\` value should render as an <a> element 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        class="px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan inline-flex items-center justify-center text-center font-medium leading-6 whitespace-nowrap select-none cursor-pointer focus:outline-none focus:ring focus:ring-opacity-60 disabled:opacity-50 disabled:pointer-events-none transition-colors duration-200"
        href="https://example.com"
        role="button"
      >
        Hello, World!
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      class="px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan inline-flex items-center justify-center text-center font-medium leading-6 whitespace-nowrap select-none cursor-pointer focus:outline-none focus:ring focus:ring-opacity-60 disabled:opacity-50 disabled:pointer-events-none transition-colors duration-200"
      href="https://example.com"
      role="button"
    >
      Hello, World!
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Button when rendered with text should display the text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan inline-flex items-center justify-center text-center font-medium leading-6 whitespace-nowrap select-none cursor-pointer focus:outline-none focus:ring focus:ring-opacity-60 disabled:opacity-50 disabled:pointer-events-none transition-colors duration-200"
        role="button"
        type="button"
      >
        Hello, World!
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="px-4 py-2 rounded text-base bg-cyan text-gray-700 shadow hover:bg-cyan-400 focus:bg-cyan-400 focus:ring-cyan inline-flex items-center justify-center text-center font-medium leading-6 whitespace-nowrap select-none cursor-pointer focus:outline-none focus:ring focus:ring-opacity-60 disabled:opacity-50 disabled:pointer-events-none transition-colors duration-200"
      role="button"
      type="button"
    >
      Hello, World!
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
